File: java/Samples/Devices/VB/DevicesSampleVB/CameraForm.vb

Imports System
Imports Neurotec.Devices
Imports Neurotec.Images
Imports System.Drawing
Imports Neurotec.Gui
Imports System.Windows.Forms

Partial Public Class CameraForm
	Inherits CaptureDeviceForm
#Region "Private fields"

	Private _cameraStatus As NCameraStatus = NCameraStatus.None
	Private _focusRegion As Nullable(Of RectangleF)
	Private _focusPen As Pen = Pens.White

#End Region

#Region "Public constructor"

	Public Sub New()
		InitializeComponent()
		OnDeviceChanged()
		OnCameraStatusChanged()
	End Sub

#End Region

#Region "Private methods"

	Private Sub OnCameraStatusChanged()
		If _cameraStatus = NCameraStatus.None Then
			cameraStatusLabel.Text = Nothing
		Else
			cameraStatusLabel.Text = _cameraStatus.ToString()
		End If
	End Sub

#End Region

#Region "Protected methods"

	Protected Overrides Function IsValidDeviceType(ByVal value As Type) As Boolean
		Return MyBase.IsValidDeviceType(value) AndAlso GetType(NCamera).IsAssignableFrom(value)
	End Function

	Protected Overrides Sub OnDeviceChanged()
		MyBase.OnDeviceChanged()
		Dim camera As NCamera = CType(Device, NCamera)
		EnableForcedCapture = camera IsNot Nothing AndAlso camera.IsStillCaptureSupported
		focusButton.Visible = camera IsNot Nothing AndAlso camera.IsFocusSupported
		resetFocusButton.Visible = focusButton.Visible
		clickToFocusCheckBox.Visible = camera IsNot Nothing AndAlso camera.IsFocusRegionSupported
	End Sub

	Protected Overrides Sub OnStatusChanged()
		MyBase.OnStatusChanged()
		focusButton.Enabled = IsCapturing
		resetFocusButton.Enabled = focusButton.Enabled
		clickToFocusCheckBox.Enabled = IsCapturing
	End Sub

	Protected Overrides Sub OnStartingCapture()
		MyBase.OnStartingCapture()
		Dim camera As NCamera = CType(Device, NCamera)
		AddHandler camera.StillCaptured, AddressOf Device_StillCaptured
	End Sub

	Protected Overrides Sub OnFinishingCapture()
		Dim camera As NCamera = CType(Device, NCamera)
		RemoveHandler camera.StillCaptured, AddressOf Device_StillCaptured
		MyBase.OnFinishingCapture()
	End Sub

	Protected Overrides Function OnObtainSample() As Boolean
		Dim camera As NCamera = CType(Device, NCamera)
		Using image As NImage = camera.GetFrame()
			If image IsNot Nothing Then
				Dim focusRegion As Nullable(Of RectangleF) = camera.FocusRegion
				SyncLock _statusLock
					_focusRegion = focusRegion
				End SyncLock
				OnImage(image, Nothing, Nothing, False)
				Return True
			Else
				Return False
			End If
		End Using
	End Function

	Protected Overrides Sub OnCaptureFinished()
		SyncLock _statusLock
			_focusRegion = Nothing
		End SyncLock
		MyBase.OnCaptureFinished()
	End Sub

#End Region

#Region "Public properties"

	Public Property ClickToFocus() As Boolean
		Get
			Return clickToFocusCheckBox.Checked AndAlso clickToFocusCheckBox.Visible
		End Get
		Set(ByVal value As Boolean)
			clickToFocusCheckBox.Checked = value
		End Set
	End Property

#End Region

#Region "Private form events"

	Private Sub Device_StillCaptured(ByVal sender As Object, ByVal e As NCameraStillCapturedEventArgs)
		If (Not HasFinal) Then
			Using image As NImage = NImage.FromStream(e.Stream)
				OnImage(image, Nothing, Nothing, True)
			End Using
			BeginInvoke(New MethodInvoker(AddressOf OnCancelCapture))
		End If
	End Sub

#End Region

	Private Sub pictureBox_Paint(ByVal sender As Object, ByVal e As System.Windows.Forms.PaintEventArgs) Handles pictureBox.Paint
		Dim focusRegion? As RectangleF
		SyncLock _statusLock
			focusRegion = _focusRegion
		End SyncLock
		If focusRegion.HasValue Then
			Dim area As Rectangle = GetPictureArea()
			e.Graphics.DrawRectangle(_focusPen, area.X + focusRegion.Value.X * area.Width, area.Y + focusRegion.Value.Y * area.Height, focusRegion.Value.Width * area.Width, focusRegion.Value.Height * area.Height)
		End If
	End Sub

	Private Delegate Sub OnPictureBoxClickedDelegate(ByVal e As MouseEventArgs, ByVal area As Rectangle, ByVal clickToFocus As Boolean)
	Private Sub OnPictureBoxClicked(ByVal e As MouseEventArgs, ByVal area As Rectangle, ByVal clickToFocus As Boolean)
		Dim camera As NCamera = CType(Device, NCamera)
		If camera IsNot Nothing AndAlso camera.IsFocusRegionSupported Then
			Dim focusRegion? As RectangleF
			SyncLock _statusLock
				focusRegion = _focusRegion
			End SyncLock
			Dim w As Single = 0.1F
			Dim h As Single = 0.1F
			If focusRegion.HasValue Then
				w = focusRegion.Value.Width
				h = focusRegion.Value.Height
			End If
			camera.FocusRegion = New RectangleF((e.X - area.X) / CSng(area.Width) - w / 2, (e.Y - area.Y) / CSng(area.Height) - h / 2, w, h)
			If clickToFocus AndAlso camera.IsFocusSupported Then
				_cameraStatus = camera.Focus()
				BeginInvoke(New MethodInvoker(AddressOf OnCameraStatusChanged))
			End If
		End If
	End Sub

	Private Sub focusButton_Click(ByVal sender As Object, ByVal e As EventArgs) Handles focusButton.Click
		NGui.InvokeAsync(AddressOf AnonymousMethod1)
	End Sub
	Private Sub AnonymousMethod1()
		Dim camera As NCamera = CType(Device, NCamera)
		_cameraStatus = camera.Focus()
		BeginInvoke(New MethodInvoker(AddressOf OnCameraStatusChanged))
	End Sub

	Private Sub resetFocusButton_Click(ByVal sender As Object, ByVal e As EventArgs) Handles resetFocusButton.Click
		NGui.InvokeAsync(AddressOf AnonymousMethod2)
	End Sub
	Private Sub AnonymousMethod2()
		Dim camera As NCamera = CType(Device, NCamera)
		camera.ResetFocus()
		_cameraStatus = NCameraStatus.None
		BeginInvoke(New MethodInvoker(AddressOf OnCameraStatusChanged))
	End Sub

	Private Sub forceButton_Click(ByVal sender As Object, ByVal e As EventArgs) Handles forceButton.Click
		NGui.InvokeAsync(AddressOf AnonymousMethod3)
	End Sub
	Private Sub AnonymousMethod3()
		Dim camera As NCamera = CType(Device, NCamera)
		_cameraStatus = camera.CaptureStill()
		BeginInvoke(New MethodInvoker(AddressOf OnCameraStatusChanged))
	End Sub
End Class

File: java/Samples/Devices/CS/DevicesSampleCS/CameraForm.cs
using System;
using Neurotec.Devices;
using Neurotec.Images;
using System.Drawing;
using Neurotec.Gui;
using System.Windows.Forms;

namespace Neurotec.Samples
{
	public partial class CameraForm : CaptureDeviceForm
	{
		#region Private fields

		private NCameraStatus _cameraStatus = NCameraStatus.None;
		private RectangleF? _focusRegion;
		private readonly Pen _focusPen = Pens.White;

		#endregion

		#region Public constructor

		public CameraForm()
		{
			InitializeComponent();
			OnDeviceChanged();
			OnCameraStatusChanged();
		}

		#endregion

		#region Private methods

		private void OnCameraStatusChanged()
		{
			cameraStatusLabel.Text = _cameraStatus == NCameraStatus.None ? null : _cameraStatus.ToString();
		}

		#endregion

		#region Protected methods

		protected override bool IsValidDeviceType(Type value)
		{
			return base.IsValidDeviceType(value) && typeof(NCamera).IsAssignableFrom(value);
		}

		protected override sealed void OnDeviceChanged()
		{
			base.OnDeviceChanged();
			var camera = (NCamera)Device;
			EnableForcedCapture = camera != null && camera.IsStillCaptureSupported;
			resetFocusButton.Visible = focusButton.Visible = camera != null && camera.IsFocusSupported;
			clickToFocusCheckBox.Visible = camera != null && camera.IsFocusRegionSupported;
		}

		protected override void OnStatusChanged()
		{
			base.OnStatusChanged();
			resetFocusButton.Enabled = focusButton.Enabled = IsCapturing;
			clickToFocusCheckBox.Enabled = IsCapturing;
		}

		protected override void OnStartingCapture()
		{
			base.OnStartingCapture();
			var camera = (NCamera)Device;
			camera.StillCaptured += Device_StillCaptured;
		}

		protected override void OnFinishingCapture()
		{
			var camera = (NCamera)Device;
			camera.StillCaptured -= Device_StillCaptured;
			base.OnFinishingCapture();
		}

		protected override bool OnObtainSample()
		{
			var camera = (NCamera)Device;
			using (NImage image = camera.GetFrame())
			{
				if (image != null)
				{
					var focusRegion = camera.FocusRegion;
					lock (_statusLock)
					{
						_focusRegion = focusRegion;
					}
					OnImage(image, null, null, false);
					return true;
				}
				return false;
			}
		}

		protected override void OnCaptureFinished()
		{
			lock (_statusLock)
			{
				_focusRegion = null;
			}
			base.OnCaptureFinished();
		}

		#endregion

		#region Public properties

		public bool ClickToFocus
		{
			get
			{
				return clickToFocusCheckBox.Checked && clickToFocusCheckBox.Visible;
			}
			set
			{
				clickToFocusCheckBox.Checked = value;
			}
		}

		#endregion

		#region Private form events

		void Device_StillCaptured(object sender, NCameraStillCapturedEventArgs e)
		{
			if (!HasFinal)
			{
				using (NImage image = NImage.FromStream(e.Stream))
				{
					OnImage(image, null, null, true);
				}
				BeginInvoke(new MethodInvoker(OnCancelCapture));
			}
		}

		#endregion

		private void pictureBox_Paint(object sender, PaintEventArgs e)
		{
			RectangleF? focusRegion;
			lock (_statusLock)
			{
				focusRegion = _focusRegion;
			}
			if (!focusRegion.HasValue) return;

			Rectangle area = GetPictureArea();
			e.Graphics.DrawRectangle(_focusPen, area.X + focusRegion.Value.X * area.Width, area.Y + focusRegion.Value.Y * area.Height,
				focusRegion.Value.Width * area.Width, focusRegion.Value.Height * area.Height);
		}

		private void pictureBox_MouseClick(object sender, MouseEventArgs e)
		{
			Rectangle area = GetPictureArea();
			bool clickToFocus = ClickToFocus;
			NGui.InvokeAsync(delegate
				{
					var camera = (NCamera)Device;
					if (camera != null && camera.IsFocusRegionSupported)
					{
						RectangleF? focusRegion;
						lock (_statusLock)
						{
							focusRegion = _focusRegion;
						}
						float w = focusRegion.HasValue ? focusRegion.Value.Width : 0.1f;
						float h = focusRegion.HasValue ? focusRegion.Value.Height : 0.1f;
						camera.FocusRegion = new RectangleF((e.X - area.X) / (float)area.Width - w / 2, (e.Y - area.Y) / (float)area.Height - h / 2, w, h);
						if (clickToFocus && camera.IsFocusSupported)
						{
							_cameraStatus = camera.Focus();
							BeginInvoke(new MethodInvoker(OnCameraStatusChanged));
						}
					}
				});
		}

		private void focusButton_Click(object sender, EventArgs e)
		{
			NGui.InvokeAsync(delegate
				{
					var camera = (NCamera)Device;
					_cameraStatus = camera.Focus();
					BeginInvoke(new MethodInvoker(OnCameraStatusChanged));
				}
			);
		}

		private void resetFocusButton_Click(object sender, EventArgs e)
		{
			NGui.InvokeAsync(delegate
				{
					var camera = (NCamera)Device;
					camera.ResetFocus();
					_cameraStatus = NCameraStatus.None;
					BeginInvoke(new MethodInvoker(OnCameraStatusChanged));
				});
		}

		private void forceButton_Click(object sender, EventArgs e)
		{
			NGui.InvokeAsync(delegate
			{
				var camera = (NCamera) Device;
				_cameraStatus = camera.CaptureStill();
				BeginInvoke(new MethodInvoker(OnCameraStatusChanged));
			});
		}
	}
}
